### PR TITLE
add rpc::Disconnector for clients to cleanly disconnect

### LINKED ISSUE #23: RpcSystem completes with error when transport is wrapped in TLS
I'm working through how to get everything hooked up with capnp-rpc-rust and tokio-tls and I have a working [proof of concept](https://github.com/ian-p-cooke/capntls) However, while the TLS version works (the client gets the response) The server always ends the RpcSystem with an error:   
`rpc error: Error { kind: Failed, description: "connection closed via error" }`
That error isn't present in the no_tls example (in my repo).

Do you have any idea as to why wrapping the RPC connection in TLS would lead to this server error?

(also posted in [reddit](https://www.reddit.com/r/rust/comments/6ld22q/capn_proto_rust_over_tls_error/) but thought this might be better as it's not a language issue)

### PR DESCRIPTION
**Pull request by ian-p-cooke**
implement `Disconnector` future that can cleanly shutdown a client's `RpcSystem`.  Resolves #23. Example of usage [here](https://github.com/ian-p-cooke/capntls/blob/master/src/client.rs#L68).

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -286,6 +286,10 @@ impl <VatId> RpcSystem <VatId> {
         handle.add(tasks);
         result
     }
+
```
Comment (dwrensha) [inline src/lib.rs:289]: This new method should get a documentation comment.

File: src/rpc.rs [hunk 2]
```diff
@@ -1518,6 +1518,75 @@ impl <VatId> ConnectionState<VatId> {
...
 }
 
+enum DisconnectorState
+{
+    Connected,
+    Disconnecting,
+    Disconnected
+}
+
+pub struct Disconnector<VatId> where VatId: 'static {
```
Comment (dwrensha) [inline src/rpc.rs]: This new struct should get a documentation comment.

File: src/rpc.rs [hunk 3]
```diff
@@ -1518,6 +1518,75 @@ impl <VatId> ConnectionState<VatId> {
...
+                }
+            },
+            DisconnectorState::Disconnected => {
+                DisconnectorState::Disconnected
+            }
+        };
+        match self.state {
+            DisconnectorState::Connected => unreachable!(),
+            DisconnectorState::Disconnecting => {
+                ::futures::task::current().notify();
```
Comment (dwrensha) [inline src/rpc.rs:1583]: This `notify()` business seem like it could lead to some unnecessary spinning if it takes a while for `self.connection_state` to become `None`. Would it maybe make more sense to stash this `Task` with the `connection_state` and then only notify it once `connection_state` becomes `None`?

**Review by dwrensha — CHANGES_REQUESTED** [review]
Cool, thanks for working on this!

Could you also add a test for the new functionality?

File: src/rpc.rs (hunk 3 above)
Comment (ian-p-cooke) [reply src/rpc.rs:1583]: I agree it could be more efficient to only poll again when you know it's going to be disconnected.  However, after calling `disconnect` you need just one turn of the event loop to get the `connection_state` to go away so in practice it's the same.  I thought it was nice that it worked without adding state anywhere else.  If you want to store the task though, who would own it?  `RpcSystem`?  Like, an `Rc<Option<Task>>` (shared with `Disconnector`) that was along side the `RpcSystem::connection_state` that you'd call `notify` on when you set the `connection_state` to `None`?   I think I prefer to keep it as-is because it's simple and I don't think you'll busy-spin unless there's a bug in `RpcSystem::disconnect`.  Let me know if I'm misunderstanding something though.

**Comment by ian-p-cooke** [comment]
I'll work on documentation and a test case this week.

File: src/rpc.rs (hunk 3 above)
Comment (dwrensha) [reply src/rpc.rs:1583]: > Like, an Rc<Option<Task>> (shared with Disconnector) that was along side the RpcSystem::connection_state that you'd call notify on when you set the connection_state to None?

Yeah, something like that is what I had in mind. I suppose it would actually need to be a `Rc<Vec<Task>>`, in case there are multiple `Disconnector`s.

Your call on whether you want to bother with this or not.

**Comment by dwrensha** [comment]
Thanks!

File: src/rpc.rs [hunk 4]
```diff
@@ -1518,6 +1518,75 @@ impl <VatId> ConnectionState<VatId> {
...
+            None => {
+                DisconnectorState::Disconnected
+            }
+        };
+        Disconnector {
+            connection_state: connection_state,
+            state: state,
+        }
+    }
+    fn disconnect(&self) {
```
Comment (calmofthestorm) [inline src/rpc.rs:1549]: Was this perhaps intended to be public? I'm having trouble figuring out the correct way to shut down the RPC system in response to an error on the transport layer.

Comment (dwrensha) [reply src/rpc.rs:1549]: Hm... yeah, it does look to me like this was intended to be public.

> the correct way to shut down the RPC system in response to an error on the transport layer

Note that I/O errors on the underlying stream automatically trigger a disconnect and shut down the RPC system. If the I/O error is [something other than a `capnp::ErrorKind::Disconnected`](https://github.com/capnproto/capnproto-rust/blob/4e6859c74251897a44c222e9679eea74145f2e33/capnp-rpc/src/lib.rs#L196), then that error gets propagated to the driving `RpcSystem` future.

Do you have a situation where that's not true, or where that's not sufficient?

(The `RpcSystem` setup and teardown logic is kinda janky right now, so I wouldn't be surprised if it has some problems.)

Comment (calmofthestorm) [reply src/rpc.rs:1549]: I'm not sure my current setup is something that's a priority to support:-)I have a TcpStream that talks to the network, and created two closures to implement a simple message format (16 bit length + payload) on top that, then encrypt/decrypt that using [Snow](https://github.com/mcginty/snow).

To give capnproto an AsyncRead/AsyncWrite, I create a UnixStream::pair with one end being given to capnproto and the other split to the encrypt/decrypt closures. I'd like to migrate this to implementing my own AsycnRead/AsyncWrite to avoid unnecessary trips to kernel space and platform-specific APIs, but that proved more complicated than I'd expected so I've been putting it off, and this mostly works except shutdown is challenging. The concept is similar to Tokio's frame codecs (which I also struggled with on a previous project some time ago).

I've used this trick a fair bit in Rust and other languages, and to get a clean shutdown, you call shutdown on all sockets (both ends of the pair and the network-facing socket) whenever there's an error detected on any socket. The problem I'm running into with Rust async_std is that it is less tolerant of my unsafely cloning the raw socket and then shutting it down from outside the executor in another thread, and panics when it can't remove the I/O watch which is honestly pretty understandable.

Anyway the right answer is for me to stop abusing Unix-specific syscalls to get around Rust ownership and async framework design, and implement AsyncRead/AsyncWrite or try tokio codecs again.

It would also have been slightly useful on a previous project (that I ended up not using capn on) to be able to cleanly shutdown the RPC system and get the socket back, but it's not clear to me how widely useful that would be, and the current disconnect API doesn't return the socket. My use case there was zero copy file transfer of out-of-band data via sendfile/splice, but again, the better solution is separate sockets for data and control plane.

So basically I can't argue that I have **good** reasons, but I do have reasons.

Comment (ian-p-cooke) [reply src/rpc.rs:1549]: I didn't make it public because it was intended to run as a Future after you were done with the connection.  see https://github.com/ian-p-cooke/capntls/blob/9b5eb98607184a05ef309203dc2ab98d352ac124/src/client.rs#L64 for my example.  That doesn't mean it -has- to be private I just didn't need it to be public for what I imagined.

Comment (calmofthestorm) [reply src/rpc.rs:1549]: Thanks -- running it as a future makes sense, and should solve my problem.
